src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php

```diff
@@ -1553,6 +1554,18 @@ private function registerMessengerConfiguration(array $config, ContainerBuilder
                 ->addTag('messenger.receiver', array('alias' => $name))
                 ->addTag('messenger.sender', array('alias' => $name))
             ;
+
+            if ('symfony://kernel.terminate' === $transport['dsn']) {
+                if (!$container->hasDefinition($transport['message_bus'])) {
+                    throw new InvalidArgumentException(sprintf('No bus was found with name: %s', $transport['message_bus']));
+                }
+
+                $container->setAlias('messenger.kernel_terminate.bus', $transport['message_bus']);
```

Comment:
I guess it rather be like:

```diff
diff --git a/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php b/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php
index 1ab2ac8a1c..3299642f5a 100644
--- a/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php
+++ b/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php
@@ -39,6 +39,7 @@ use Symfony\Component\Console\Command\Command;
 use Symfony\Component\DependencyInjection\Alias;
 use Symfony\Component\DependencyInjection\Argument\ServiceClosureArgument;
 use Symfony\Component\DependencyInjection\ChildDefinition;
+use Symfony\Component\DependencyInjection\Compiler\ServiceLocatorTagPass;
 use Symfony\Component\DependencyInjection\ContainerBuilder;
 use Symfony\Component\DependencyInjection\ContainerInterface;
 use Symfony\Component\DependencyInjection\Definition;
@@ -1518,6 +1519,7 @@ class FrameworkExtension extends Extension
             'before' => array(array('id' => 'logging')),
             'after' => array(array('id' => 'route_messages'), array('id' => 'call_message_handler')),
         );
+        $buses = array();
         foreach ($config['buses'] as $busId => $bus) {
             $middleware = $bus['default_middleware'] ? array_merge($defaultMiddleware['before'], $bus['middleware'], $defaultMiddleware['after']) : $bus['middleware'];
 
@@ -1529,6 +1531,7 @@ class FrameworkExtension extends Extension
 
             $container->setParameter($busId.'.middleware', $middleware);
             $container->register($busId, MessageBus::class)->addArgument(array())->addTag('messenger.bus');
+            $buses[$busId] = new Reference($busId);
 
             if ($busId === $config['default_bus']) {
                 $container->setAlias('message_bus', $busId)->setPublic(true);
@@ -1538,6 +1541,10 @@ class FrameworkExtension extends Extension
             }
         }
 
+        $container->getDefinition('messenger.transport.kernel_terminate.factory')
+            ->replaceArgument(0, ServiceLocatorTagPass::register($container, $buses))
+        ;
+
         if (!$container->hasAlias('message_bus')) {
             throw new LogicException(sprintf('The default bus named "%s" is not defined. Define it or change the default bus name.', $config['default_bus']));
         }
@@ -1556,11 +1563,13 @@ class FrameworkExtension extends Extension
             ;
 
             if ('symfony://kernel.terminate' === $transport['dsn']) {
-                if (!$container->hasDefinition($transport['message_bus'])) {
-                    throw new InvalidArgumentException(sprintf('No bus was found with name: %s', $transport['message_bus']));
+                $busId = $transport['options']['bus'] ?? null;
+                if (null === $busId) {
+                    throw new InvalidArgumentException(sprintf('Missing mandatory "bus" option for transport "%s"', $name));
+                }
+                if (!$container->hasDefinition($busId)) {
+                    throw new InvalidArgumentException(sprintf('No bus was found with id: %s', $busId));
                 }
-
-                $container->setAlias('messenger.kernel_terminate.bus', $transport['message_bus']);
 
                 $transportDefinition->addTag('kernel.event_listener', array('event' => KernelEvents::TERMINATE, 'method' => 'onKernelTerminate'));
                 $transportDefinition->addTag('kernel.event_listener', array('event' => KernelEvents::EXCEPTION, 'method' => 'onKernelException'));
diff --git a/src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger.xml b/src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger.xml
index ce936d13d1..76ecb3821a 100644
--- a/src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger.xml
+++ b/src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger.xml
@@ -64,7 +64,7 @@
         </service>
 
         <service id="messenger.transport.kernel_terminate.factory" class="Symfony\Component\Messenger\Transport\KernelTerminate\SymfonyKernelEventTransportFactory">
-            <argument type="service" id="messenger.kernel_terminate.bus" />
+            <argument /> <!-- Message bus locator -->
             <tag name="messenger.transport_factory" />
         </service>
     </services>
diff --git a/src/Symfony/Component/Messenger/Transport/KernelTerminate/SymfonyKernelEventTransportFactory.php b/src/Symfony/Component/Messenger/Transport/KernelTerminate/SymfonyKernelEventTransportFactory.php
index 343f8d2906..f76a773ea0 100644
--- a/src/Symfony/Component/Messenger/Transport/KernelTerminate/SymfonyKernelEventTransportFactory.php
+++ b/src/Symfony/Component/Messenger/Transport/KernelTerminate/SymfonyKernelEventTransportFactory.php
@@ -11,23 +11,23 @@
 
 namespace Symfony\Component\Messenger\Transport\KernelTerminate;
 
-use Symfony\Component\Messenger\MessageBusInterface;
+use Psr\Container\ContainerInterface;
 use Symfony\Component\Messenger\Transport\MemoryTransport;
 use Symfony\Component\Messenger\Transport\TransportFactoryInterface;
 use Symfony\Component\Messenger\Transport\TransportInterface;
 
 class SymfonyKernelEventTransportFactory implements TransportFactoryInterface
 {
-    private $bus;
+    private $busLocator;
 
-    public function __construct(MessageBusInterface $bus)
+    public function __construct(ContainerInterface $busLocator)
     {
-        $this->bus = $bus;
+        $this->busLocator = $busLocator;
     }
 
     public function createTransport(string $dsn, array $options): TransportInterface
     {
-        return new MemoryTransport($this->bus);
+        return new MemoryTransport($this->busLocator->get($options['bus']));
     }
 
     public function supports(string $dsn, array $options): bool
```

Reply:
Let's see what others say, but I think my version is simpler.